doc/design/attribute.md

```diff
@@ -0,0 +1,123 @@
+
+* Caffe2 uses `proto2` syntax. It treats all attributes as a list, and each attribute contains a `name`. Each time caffe2 read an attribute is searching a variable in a list. It is slow if the number of attributes is large. Caffe2 also mark all field as `optional`. It doesn't ensure `one of` attribute value is set.
+* By using `proto3` syntax in tensorflow, the attribute implementation in tensorflow is using `map`, and `oneof` keywords. Looking up from attribute map in tensorflow is fast.
+
+Paddle is using `protobuf 3` as its dependency library. By simplify `tensorflow`'s implementation, Paddle's Attribute protobuf message schema could be
+
+```protobuf
+message Attribute {
+    message ListValue {
+        repeated int32 ints = 1;
+        repeated float floats = 2;
+        repeated string strings = 3;
+    }
+
+    oneof value {
+        ListValue list = 1;
+        int32 i = 2;
+        float f = 3;
+        string s = 4;
+    }
+}
+```
+
+In `OperatorDescription` message, there should be a field like this:
+
+```protobuf
+message OperatorDescription {
+  map<string, Attribute> attrs;
+}
+```
+
+## CPP implementation
+
+### AttributeReader
+
+In CPP, it should be a helper class for reading `map<string, Attribute>`. The reading method should accept a template parameter, which is the type of Attribute.  If type mismatch or attribute is not found, `Get` method should return an `Error`. That helper class we named `AttributeReader`.
+
+The interface of `AttributeReader` is like this:
+
+```cpp
+using AttributeMap = google::protobuf::Map<std::string, Attribute>;
+class AttributeReader {
+ public:
+  explicit AttributeReader(const AttributeMap& attrs) : attrs_(attrs) {}
+
+  template <typename T>
+  Error __must_check Get(const std::string& attributeName, T* attr) const;
+
+  template <typename T>
+  Error __must_check GetArray(const std::string& attributeName,
+                              std::vector<T>* array) const;
+
+ private:
+  const AttributeMap& attrs_;
+};
+```
+
+There are two methods in `AttributeReader`: `Get` and `GetArray`. `GetArray` is used for `ListValue`, and `Get` is used for the rests. The user should invoke either of them when he wants to get an Attribute value from `AttributeMap`.
+
+### Attribute in Operator
+
+Each operator stores its attributes. For faster attribute access, we should not let user parse `AttributeMap` during `Run` method in Operator. When `NetworkBase` adds an operator to computation graph, the `Attribute` could be parsed, and stored in each operator's the private member.
+
+```cpp
+class OperatorBase {
+ public:
+  Error InitializeAttribute(const AttributeReader& attrs) = 0;
+};
+
+class CosineOp : public OperatorBase {
+ public:
+  Error InitializeAttribute(const AttributeReader& attrs) {
+    auto err = attrs.Get<float>("scale", &scale_);
+    if (!err.isOK() && err != "Attribute Not Found") {
```

Comment:
If the error is `Attribute Not Found`, then the default value is used.

The default value is set in private data member, like `float scale_{1.0};`.